### PR TITLE
Introduce Edge Agent App Settings

### PR DESCRIPTION
**Pull request by philipktlin**
Introduce EdgeAgentAppSettings in order to:
1. Centralize 1 single place to access .Net Configuration (except DockerRegistryAuthConfigSection which plan to do it later to minimize changes in this PR).  This will help for testing.
2. Strong type class to auto-mapping to fields in appsettings json; which helps to do conversion.
3. Add business validation for parameters right after loading appsettings json; so fail as earlier as possible.

Note: this is 1st version of EdgeAgentAppSettings, I will continue to refactor it; for example, group parameters by runtime type in 2 subclasses, etc.

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/Constants.cs [hunk 1]
```diff
@@ -17,9 +14,7 @@ public static class Constants
         public const string ModuleIdKey = "ModuleId";
 
         public const RestartPolicy DefaultRestartPolicy = RestartPolicy.OnUnhealthy;
-
-        public const ModuleStatus DefaultDesiredStatus = ModuleStatus.Running;
-
+
```
Comment (arsing) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/Constants.cs]: Trailing whitespace on this line, and two more cases below.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/ConfigSource.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+namespace Microsoft.Azure.Devices.Edge.Agent.Core
+{
+    public enum ConfigSource
+    {
+        Local,
+        Twin
+    }
+}
```
Comment (arsing) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/ConfigSource.cs]: Missing newline at EOF.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/EdgeRuntimeMode.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+namespace Microsoft.Azure.Devices.Edge.Agent.Core
+{
+    public enum EdgeRuntimeMode
+    {
+        Iotedged,
+        Docker
+    }
+}
```
Comment (arsing) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/EdgeRuntimeMode.cs]: Missing newline at EOF.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerLoggingConfig.cs [hunk 4]
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) Microsoft. All rights reserved.
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (arsing) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerLoggingConfig.cs]: Blank line after copyright header like the other files?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/TwinConfigSourceModule.cs [hunk 5]
```diff
@@ -14,26 +14,17 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Service.Modules
     using Microsoft.Azure.Devices.Edge.Agent.IoTHub.ConfigSources;
     using Microsoft.Azure.Devices.Edge.Agent.IoTHub.Reporters;
     using Microsoft.Azure.Devices.Edge.Util;
-    using Microsoft.Extensions.Configuration;
-
+
```
Comment (arsing) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/TwinConfigSourceModule.cs]: Trailing whitespace.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/DependencyManager.cs [hunk 6]
```diff
@@ -109,7 +109,7 @@ void RegisterMqttModule(ContainerBuilder builder, (bool isEnabled, bool usePersi
 
         void RegisterRoutingModule(ContainerBuilder builder, (bool isEnabled, bool usePersistentStorage, StoreAndForwardConfiguration config, string storagePath) storeAndForward)
         {
-            var routes = this.configuration.GetSection("routes").Get<Dictionary<string, string>>();
+            Dictionary<string, string> routes = this.configuration.GetSection("routes").Get<Dictionary<string, string>>();
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/DependencyManager.cs]: It cannot resolve type information from generic type argument.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/Constants.cs [hunk 7]
```diff
@@ -71,8 +58,6 @@ public static class Constants
 
         public const string IotedgedMode = "iotedged";
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/Constants.cs]: IotedgeMode is used in CreateOrUpdateCommand class line 137.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/IAgentAppSettings.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+namespace Microsoft.Azure.Devices.Edge.Agent.Core
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Net;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.Logging;
+
+    public interface IAgentAppSettings
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/IAgentAppSettings.cs]: Just do minimum for this PR; we can further refactor later on.  Will add comments here.

[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/IAgentAppSettings.cs]: Yes, with the interface, it makes mocking much easier in tests.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/RestartPolicyManager.cs [hunk 9]
```diff
@@ -60,7 +62,7 @@ public ModuleStatus ComputeModuleStatusFromRestartPolicy(ModuleStatus status, Re
             modules.Where(module => this.ShouldRestart(module));
 
         internal TimeSpan GetCoolOffPeriod(int restartCount) =>
-            TimeSpan.FromSeconds(Math.Min(this.coolOffTimeUnitInSeconds * Math.Pow(2, restartCount), MaxCoolOffPeriodSecs));
+            TimeSpan.FromSeconds(Math.Min(this.coolOffTimeUnit.Seconds * Math.Pow(2, restartCount), MaxCoolOffPeriodSecs));
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/RestartPolicyManager.cs]: good catch. thanks.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/AgentAppSettings.cs [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+            this.ConfigRefreshFrequency = TimeSpan.FromSeconds(Convert.ToInt32(this.appSettings.ConfigRefreshFrequencySecs ?? "3600"));
+            this.DockerRegistryAuthConfigSection = config.GetSection("DockerRegistryAuth");
+            this.HttpsProxy = Proxy.Parse(this.appSettings.Https_proxy, this.Logger);
+            this.StoragePath = this.GetStoragePath(this.appSettings.StorageFolder);
+            this.UpstreamProtocol = this.appSettings.UpstreamProtocol.ToUpstreamProtocol();
+            this.VersionInfo = VersionInfo.Get(VersionInfoFileName);
+        }
+
+        void SetupLogger()
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/AgentAppSettings.cs]: It is not easy to test on MainAsync; therefore I moved in this class and btw SetupLogger is 1st thing to call in MainAsync through AgentAppSettings.Initialize method.  Let me know if you still have any concerns.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/Microsoft.Azure.Devices.Edge.Agent.IoTHub.csproj [hunk 11]
```diff
@@ -22,6 +22,8 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.Azure.Devices" Version="1.17.3" />
     <PackageReference Include="Microsoft.Azure.Devices.Client" Version="1.21.0" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="2.1.1" />
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/Microsoft.Azure.Devices.Edge.Agent.IoTHub.csproj]: removed.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/AgentAppSettings.cs [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+            Preconditions.CheckRange(this.MaxRestartCount, 1, nameof(this.MaxRestartCount));
+            Preconditions.CheckNonWhiteSpace(this.StoragePath, nameof(this.StoragePath));
+            Preconditions.CheckNotNull(this.VersionInfo, nameof(this.VersionInfo));
+        }
+
+        // All these property names should be case-insensitive matched to field name in AppSettings json file or name of environment variable.
+        // Don't remove set method of these properties; it is used when configuration binding.
+        class AppSettings
+        {
+            public string BackupConfigFilePath { get; set; }
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/AgentAppSettings.cs]: This class is used to map app setting json by out-of-the-box config.get<T> method; which will simplify a little bit; e.g. DockerLoggingOptions is a dictionary; we don't need to write any code to handle it.  Should it be fine to use it?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/AgentAppSettings.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/AgentAppSettings.cs]: Are you planning on adding tests for Main?
In any case, I don't understand the concern of moving logger setup to main. If something fails here, we have no logs.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/AgentAppSettings.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/AgentAppSettings.cs]: I understand the intent.. it is fine I guess.

**Comment by philipktlin** [comment]
Will create a new PR since this branch is out-of-date and difficult to merge with latest changes.  Will re-implement again and try to break it down in multiple PRs.

**Pull request closed without merging**